tftp/util.py

```
@@ -2,126 +2,73 @@
-    Optionally calls a timeout handler, if provided, when there are no more timeout values.
+    The given C{timings} is an iterable of numbers. Each is a delay in seconds
+    that will be taken before making the next call to C{call} or C{last}.
 
-    @param timeout: an iterable, that yields valid _seconds arguments to
-    L{callLater<twisted.internet.interfaces.IReactorTime.callLater>} (floats)
-    @type timeout: any iterable
+    The call to C{last} will happen after the last delay. If C{timings} is an
+    infinite iterable then C{last} will never be called.
 
-    @param run_now: whether or not the callable should be called immediately
-    upon initialization. Relinquishes control to the reactor
-    (calls callLater(0,...)). Default: C{False}.
-    @type run_now: C{bool}
-
-    @param callable: the callable, that will be called at the specified intervals
-    @param callable_args: the arguments to call it with
-    @param callable_kwargs: the keyword arguments to call it with
-
-    @param on_timeout: the callable, that will be called when there are no more
-    timeout values
-    @param on_timeout_args: the arguments to call it with
-    @param on_timeout_kwargs: the keyword arguments to call it with 
-    
+    This returns a C{Deferred} which can be cancelled. If the cancellation is
+    successful — i.e. there is something to cancel — then the result is set to
+    L{CANCELLED}. In other words, L{CancelledError} is squashed into a
+    non-failure condition.
     """
+    timings = iterlast(timings)
+
+    def iterate(_=None):
+        for is_last, delay in timings:
+            # Return immediately with a deferred call.
+            if is_last:
+                return deferLater(clock, delay, last)
+            else:
+                return deferLater(clock, delay, call).addCallback(iterate)
+        else:
+            # No timings were given.
+            return succeed(None)
```

Feel like either no timings should raise an error or should just call `last` instantly (but that is some what of a weird behavior).

Strange that you have it just resolving to None.